File: code/task3train.py
from collections import defaultdict
from itertools import combinations
from pyspark import SparkContext
from math import sqrt
import json
import random
import sys
import time


class T3t:
    def __init__(self) -> None:
        self.trainfile = sys.argv[1]
        self.outmodelfile = sys.argv[2]
        self.cf_type = sys.argv[3]

    @staticmethod
    def gen_hash_fns(num_funcs):
        step = 10 ** 7
        curstep = step
        start = 10 ** 9 - 10 ** 8 + 17
        fns = [None] * num_funcs

        for i in range(num_funcs):
            a = random.randint(start, start + curstep - 10 ** 7 + 23)
            b = random.randint(10 ** 9 - 10 ** 8 + 30697,
                               10 ** 9 + curstep + 10 ** 8 - 4123892)
            curstep += step
            fns[i] = (a, b)
        return fns

    @staticmethod
    def gen_signatures(uid_values, hash_params, num_buckets):
        p = 10 ** 9 + 7
        sig = []
        for a, b in hash_params:
            sig_row = []
            for v in uid_values:
                sig_row.append(((a * v + b) % p) % num_buckets)
            sig.append(min(sig_row))
        return sig

    @staticmethod
    def gen_candidate_pairs(sig_temp, num_bands):
        candidate_pairs = set()
        for i in range(num_bands):
            cur_bucket = defaultdict(set)
            for row in sig_temp:
                cur_bucket[row[1][i]].add(row[0])
            for v in cur_bucket.values():
                if len(v) == 1:
                    continue
                for b1, b2 in combinations(sorted(v), 2):                    
                    candidate_pairs.add((b1, b2))
        return candidate_pairs

    @staticmethod
    def get_ps(rating1, rating2):
        k1set = set(rating1.keys())
        k2set = set(rating2.keys())
        intsc = k1set.intersection(k2set)

        c1 = []
        c2 = []
        for c in intsc:
            c1.append(rating1[c])
            c2.append(rating2[c])
        mean1 = (sum(c1) / len(intsc))
        mean2 = (sum(c2) / len(intsc))
        rbaru = [-mean1 + r for r in c1]
        rbarv = [-mean2 + r for r in c2]
        num = sum([rbaru[i] * rbarv[i] for i in range(len(rbaru))])
        bot = sqrt(sum([pow(x, 2) for x in rbaru])) * sqrt(sum([pow(x, 2) for x in rbarv]))
        return 0 if not (num > 0 and bot > 0) else num / bot

    @staticmethod
    def get_actual_pairs(candidates, ubr):
        actual_similar_bizz = []
        print("Number of candidate pairs: {}".format(len(candidates)))
        for u1, u2 in candidates:
            rating1 = ubr[u1]
            rating2 = ubr[u2]
            k1set = set(rating1.keys())
            k2set = set(rating2.keys())

            intsc = k1set & k2set
            if len(intsc) >= 3:
                sim = len(intsc) / len(k1set | k2set)
                if sim >= 0.01:
                    ps = T3t.get_ps(rating1, rating2)
                    if ps > 0:
                        actual_similar_bizz.append((u1, u2, ps))
        return actual_similar_bizz

    @staticmethod
    def gen_burmap(burmap):
        ret = defaultdict(dict)

        for biz_id, value in burmap.items():
            for u, r in value:
                ret[biz_id][u] = r
        return ret

    def run_user_based(self):
        st = time.time()
        sc = SparkContext.getOrCreate()
        sc.setLogLevel("OFF")

        num_bands = 70
        num_hashes = 70
        textRDD = sc.textFile(self.trainfile).map(lambda row: json.loads(row))
        textRDD.cache()

        bizRDD = textRDD.map(lambda row: row["business_id"]).distinct()
        cmp_map = bizRDD.zipWithIndex().collectAsMap()
        num_buckets = len(cmp_map) - 1
        hash_params = T3t.gen_hash_fns(num_hashes)

        # (user_id, {biz1, biz2, ...})
        biz_sets = textRDD.map(lambda row: (row['user_id'], cmp_map[row['business_id']])).distinct(
        ).groupByKey().map(lambda row: (row[0], set(row[1])))
        biz_sets.cache()

        sig_temp = biz_sets.mapValues(lambda uids_list: T3t.gen_signatures(uids_list, hash_params, num_buckets)).collect()
        
        candidate_pairs = T3t.gen_candidate_pairs(sig_temp, num_bands)

        ubRDD = textRDD.map(lambda row: (row["user_id"], [(row["business_id"], row["stars"])])).reduceByKey(lambda x, y: x + y).collect()
        ubr = {row[0] : {x[0] : x[1] for x in row[1]} for row in ubRDD}
        print("Running time before actual pairs: {}".format(time.time() - st))
        actual_pairs = T3t.get_actual_pairs(candidate_pairs, ubr)
        
        with open(self.outmodelfile, "w+") as f:
            for useru, userv, sim in actual_pairs:
                f.write(json.dumps({"u1": useru, "u2": userv, "sim": sim}) + "\n")

    def run_item_based(self):
        sc = SparkContext.getOrCreate()
        sc.setLogLevel("OFF")

        ubrRDD = sc.textFile(self.trainfile).map(lambda row: json.loads(row)).map(lambda row: (row["business_id"], (row["user_id"], row["stars"])))
        burMap = T3t.gen_burmap(ubrRDD.groupByKey().map(lambda row: (row[0], list(row[1]))).filter(lambda row: len(row[1]) >= 3).collectAsMap())
        unique_biz = burMap.keys()

        similar_bizz = []
        for biz1, biz2 in combinations(unique_biz, 2):
            rating1 = burMap[biz1]
            rating2 = burMap[biz2]

            intsc = set(rating1.keys()).intersection(set(rating2.keys()))
            if len(intsc) > 2:
                ps = T3t.get_ps(rating1, rating2)
                if ps > 0:
                    similar_bizz.append((biz1, biz2, ps))
        
        with open(self.outmodelfile, "w+") as f:
            for biz1, biz2, ps in similar_bizz:
                f.write(json.dumps({"b1": biz1, "b2": biz2, "sim": ps}) + "\n")

    def run(self):
        if self.cf_type == "user_based":
            self.run_user_based()
        else:
            self.run_item_based()

if __name__ == "__main__":
    t3 = T3t()

    st = time.time()
    t3.run()
    et = time.time()
    print("Total time taken: {}".format(et - st))

File: hw3/src/main/scala/task3train.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD
import org.apache.spark.storage.StorageLevel
import org.json4s._
import org.json4s.jackson.JsonMethods._

import scala.math.log10
import scala.util.control.Breaks._
import scala.util.Random._
import collection.mutable.{ArrayBuffer, HashMap, HashSet}
import java.io.{File, PrintWriter}
import scala.io.Source
import scala.math.Ordering.Implicits._
import scala.math.sqrt
import scala.util.Random


object task3train {
  implicit val formats = DefaultFormats

  def genHashFns(numFuncs: Int): Array[(Long, Long)] = {
    val start = (10e7 - 5128423).toInt
    val end = (10e8 + 3267231).toInt
    val bstart = (10e7 - 3069127).toInt
    val bend = (10e8 + 1278127).toInt
    var fns = new ArrayBuffer[(Long, Long)]
    val random = new Random()

    for (i <- 0 to numFuncs) {
      val a = start + random.nextInt(end - start)
      val b = bstart + random.nextInt(bend - bstart)
      fns += Tuple2(a, b)
    }

    return fns.toArray
  }

  def genSignatures(uidValues: Set[Long], hashParams: Array[(Long, Long)], numBuckets: Int): Array[Long] = {
    val p = (10e9 + 7).toInt
    var sig = new ArrayBuffer[Long]

    for (params <- hashParams) {
      val a = params._1
      val b = params._2
      var sig_row = new ArrayBuffer[Long]

      for (v <- uidValues) {
        sig_row += (((a * v + b) % p) % numBuckets)
      }

      sig += sig_row.min
    }

    return sig.toArray
  }

  def loadRow(row: JValue): (String, (String, Double)) = {
    val s1 = (row\"user_id").extract[String]
    val t1 = (row\"business_id").extract[String]
    val t2 = (row\"stars").extract[Double]

    return (s1, (t1, t2))
  }

  def genUBR(ubRDD: Array[(String, Array[(String, Double)])]): Map[String, HashMap[String, Double]] = {
    val ubr = new HashMap[String, HashMap[String, Double]]
    for (row <- ubRDD) {
      val tempMap = new HashMap[String, Double]
      for (x <- row._2) {
        tempMap(x._1) = x._2
      }
      ubr(row._1) = tempMap
    }
    return ubr.toMap
  }

  def pearsonSim(r1: HashMap[String, Double], r2: HashMap[String, Double]): Double = {
    val k1set = r1.keys.toSet
    val k2set = r2.keys.toSet
    val intsc = k1set.intersect(k2set)

    val c1 = new ArrayBuffer[Double]
    val c2 = new ArrayBuffer[Double]

    for (c <- intsc) {
      c1 += r1(c)
      c2 += r2(c)
    }

    val mean1 = c1.sum.toDouble / c1.size.toDouble
    val mean2 = c2.sum.toDouble / c2.size.toDouble

    val v1 = new ArrayBuffer[Double]
    val v2 = new ArrayBuffer[Double]

    for (r <- c1) {
      v1 += (r - mean1)
    }

    for (r <- c2) {
      v2 += (r - mean2)
    }

    var num = 0.0
    var den = 0.0

    for (i <- 0 to v1.size) {
      num += (v1(i) * v2(i))
    }

    den = sqrt(v1.map(x => (x * x)).sum) * sqrt(v2.map(x => (x * x)).sum)
    if (num > 0 && den > 0) {
      return num / den
    }
    return 0
  }

  def writeData(simUsers: Array[(String, String, Double)], outputfile: String): Unit = {
    val out = new PrintWriter(new File(outputfile))

    for (item <- simUsers) {
      val output: Map[String, Any] = Map("u1"->item._1, "u2"->item._2, "sim"->item._3)
      val formatted_output = org.json4s.jackson.Serialization.write(output)
      out.write(formatted_output)
      out.write("\n")
    }

    out.close()
  }

  def main(args: Array[String]) = {
    val t1 = System.nanoTime

    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("OFF")
    val trainfile = args(0)
    val modelfile = args(1)
    val cfType = args(2)

    val NUM_BANDS = 50
    val NUM_HASHES = 50

    val textRDD = sc.textFile(trainfile).map(row => parse(row))

    val bizRDD = textRDD.map(row => (row\"business_id").extract[String]).distinct()
    val cmpMap = bizRDD.zipWithIndex().collectAsMap()
    val numBuckets = cmpMap.size - 1
    val hashParams = genHashFns(NUM_HASHES)

    val bizSets = textRDD.map(row => ((row \ "user_id").extract[String], cmpMap((row \ "business_id").extract[String])))
      .distinct()
      .groupByKey()
      .map(row => (row._1, row._2.toSet))
    bizSets.cache()
    val bizMap = bizSets.collectAsMap()

    println("Done generating bizSets and bizMap")

    val sigTemp = bizSets.mapValues(uidsList => genSignatures(uidsList, hashParams, numBuckets)).collect()

    println("Done generating sigTemp")

    var cands = new HashSet[(String, String)]()

    for (i <- 0 to NUM_BANDS) {
      var curBucket = new HashMap[Long, HashSet[String]].withDefaultValue(HashSet[String]())
      for (row <- sigTemp) {
        if (curBucket.contains(row._2(i))) {
          curBucket(row._2(i)) += row._1
        } else {
          curBucket(row._2(i)) = new HashSet[String]
          curBucket(row._2(i)) += row._1
        }
      }
      for (v <- curBucket.values) {
        if (v.size > 1) {
          for (comb <- v.toArray.sorted.combinations(2)) {
            cands += Tuple2(comb(0), comb(1))
          }
        }
      }
    }

    println("Done generating Candidates")
    println(cands.size)

    val ubRDD = textRDD
      .map(row => loadRow(row))
      .groupByKey()
      .mapValues(x => x.toArray)
      .collect()

    val ubr = genUBR(ubRDD)

    var actualSimBizz = new ArrayBuffer[(String, String, Double)]
    for (pair <- cands) {
      val u1 = pair._1
      val u2 = pair._2

      val rating1 = ubr(u1)
      val rating2 = ubr(u2)

      val k1set = rating1.keys.toSet
      val k2set = rating2.keys.toSet

      val intsc = k1set.intersect(k2set)

      if (intsc.size >= 3) {
        val sim = intsc.size / (k1set.union(k2set)).size
        if (sim >= 0.01) {
          val ps = pearsonSim(rating1, rating2)
          if (ps > 0)
            actualSimBizz += Tuple3(u1, u2, sim)
        }
      }
    }

    println("Done generating similar pairs")

    writeData(actualSimBizz.toArray, outputfile = modelfile)

    println("Duration: " + (System.nanoTime - t1) / 1e9d)
  }
}